doc/examples/denoise_ascm.py

```diff
@@ -0,0 +1,179 @@
+"""
+==============================================================
+Denoise images using Adaptive Soft Coefficient Matching (ASCM)
+==============================================================
+
+Using the non-local means based adaptive denoising [Coupe11]_ you can denoise
+3D or 4D images and boost the SNR of your datasets.
+
+The adaptive soft coefficient matching (ASCM) takes in the noisy input image
+and two denoised versions of the noisy input, and gives a better denoised
+image as the output. Here, one denoised input is more "smooth" than the other
+(the easiest way to achieve this denoising is use ``non_local_means`` with
+two different patch radii).
+
+ASCM involves these basic steps
+
+* Computes wavelet decomposition of the noisy as well as denoised inputs
+
+* Combines the wavelets for the output image in a way that it takes it's
+  smoothness (low frequency components) from the input with larger smoothing,
+  and the sharp features (high frequency components) from the input with
+  less smoothing.
+
+This way ASCM gives us a well denoised output while preserving the sharpness
+of the image features, which is the primary goal in any denoising algorithm.
+
+Let us load the necessary modules
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import nibabel as nib
+from dipy.data import (fetch_sherbrooke_3shell,
+                       read_sherbrooke_3shell)
+from dipy.denoise.noise_estimate import estimate_sigma
+from time import time
+from dipy.denoise.non_local_means import non_local_means
+from dipy.denoise.adaptive_soft_matching import adaptive_soft_matching
+
+"""
+Choose one of the data from the datasets in DIPY
+"""
+
+fetch_sherbrooke_3shell()
+img, gtab = read_sherbrooke_3shell()
+
+data = img.get_data()
+affine = img.get_affine()
+
+mask = data[..., 0] > 80
+data = data[..., 1]
+
+print("vol size", data.shape)
+
+t = time()
+
+"""
+The ``ascm`` function takes two denoised inputs, one more smooth than the
+other, for generating these inputs we will use the ``non_local_means``
+denoising. In order to call ``non_local_means`` first you need to estimate
```

Comment:
You can just say: 'In order to compute the two required pre-denoised versions of the data, first you need to estimate the ...'